From 44c09f6c4da639674dc9fa4b3dfa5c7d752cdf3e Mon Sep 17 00:00:00 2001
From: Dave Airlie <airlied@linux.ie>
Date: Tue, 7 Oct 2008 02:38:44 +1000
Subject: [PATCH 32/36] exa: don't call composite routines with no buffer.

We can get a case with gnome-terminal + links, where we get two arrays
of glyphs all with 0 width and 0 heights in them. If this happens
we manage to get to this case without any buffer setup and segfault.
(cherry picked from commit 717c7492a0f6ba3fb3eabda33515881eef314155)
---
 exa/exa_glyphs.c |   14 ++++++++------
 1 files changed, 8 insertions(+), 6 deletions(-)

diff --git a/exa/exa_glyphs.c b/exa/exa_glyphs.c
index 2acc891..169763f 100644
--- a/exa/exa_glyphs.c
+++ b/exa/exa_glyphs.c
@@ -871,12 +871,14 @@ exaGlyphs (CARD8 	 op,
 	list++;
     }
     
-    if (maskFormat)
-	exaGlyphsToMask(pMask, &buffer);
-    else
-	exaGlyphsToDst(op, pSrc, pDst, &buffer,
-		       xSrc, ySrc, xDst, yDst);
-    
+    if (buffer.count) {
+        if (maskFormat)
+	    exaGlyphsToMask(pMask, &buffer);
+        else
+	    exaGlyphsToDst(op, pSrc, pDst, &buffer,
+		           xSrc, ySrc, xDst, yDst);
+    }
+
     if (maskFormat)
     {
 	x = extents.x1;
-- 
1.6.0.3

